intermediate_source/pipeline_tutorial.py

```diff
@@ -0,0 +1,458 @@
+"""
+Training Transformer models using Pipeline Parallelism
+======================================================
+
+**Author**: `Pritam Damania <https://github.com/pritamdamania87>`_
+
+This tutorial demonstrates how to train a large Transformer model across
+multiple GPUs using pipeline parallelism. This tutorial is an extension of the
+`Sequence-to-Sequence Modeling with nn.Transformer and TorchText <https://pytorch.org/tutorials/beginner/transformer_tutorial.html>`__ tutorial
+and scales up the same model to demonstrate how pipeline parallelism can be
+used to train Transformer models.
+
+Prerequisites:
+
+    * `Pipeline Parallelism <https://pytorch.org/docs/stable/pipeline.html>`__
+    * `Sequence-to-Sequence Modeling with nn.Transformer and TorchText <https://pytorch.org/tutorials/beginner/transformer_tutorial.html>`__
+"""
+
+
+######################################################################
+# Define the model
+# ----------------
+#
+
+
+######################################################################
+# In this tutorial, we will split a Transformer model across two GPUs and use
+# pipeline parallelism to train the model. The model is exactly the same model
+# used in the `Sequence-to-Sequence Modeling with nn.Transformer and TorchText
+# <https://pytorch.org/tutorials/beginner/transformer_tutorial.html>`__ tutorial,
+# but is split into two stages. The largest number of parameters belong to the
+# `nn.TransformerEncoder <https://pytorch.org/docs/stable/generated/torch.nn.TransformerEncoder.html>`__ layer.
+# The `nn.TransformerEncoder <https://pytorch.org/docs/stable/generated/torch.nn.TransformerEncoder.html>`__
+# itself consists of ``nlayers`` of `nn.TransformerEncoderLayer <https://pytorch.org/docs/stable/generated/torch.nn.TransformerEncoderLayer.html>`__.
+# As a result, our focus is on ``nn.TransformerEncoder`` and we split the model
+# such that half of the ``nn.TransformerEncoderLayer`` are on one GPU and the
+# other half are on another. To do this, we pull out the ``Encoder`` and
+# ``Decoder`` sections into seperate modules and then build an nn.Sequential
+# representing the original Transformer module.
+
+import sys
+import math
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+import tempfile
+from torch.nn import TransformerEncoder, TransformerEncoderLayer
+
+if sys.platform == 'win32':
+    print('Windows platform is not supported for pipeline parallelism')
+    sys.exit(0)
+if torch.cuda.device_count() < 2:
+    print('Need at least two GPU devices for this tutorial')
+    sys.exit(0)
```

Comment:
Those kind of checks do not work nicely with tutorial building system.
Would it be possible to add them to `__main__`?
cc: @brianjo